lib/AST/TypeJoinMeet.cpp

```diff
@@ -198,5 +198,21 @@ Type Type::join(Type first, Type second) {
     return Type();
   }
 
+  // FIXME: Remove this once all of the cases are implemented.
+  // If one or both of the types are optional types,
+  // look at the underlying object type.
+  bool isFirstOptional, isSecondOptional;
+  Type objectType1 = first->getOptionalObjectType(isFirstOptional);
+  Type objectType2 = second->getOptionalObjectType(isSecondOptional);
+
+  if (isFirstOptional || isSecondOptional) {
+    // Compute the join of the unwrapped type. If there is none, we're done.
+    Type unwrappedJoin = join(objectType1 ? objectType1 : first,
+                              objectType2 ? objectType2 : second);
+    if (!unwrappedJoin) return nullptr;
```

Comment:
This isn't quite right. The old code (before any of my changes) returned nullptr from `join()` if the join of the unwrapped objects wasn't computable by the code. The current code returns `Any` in this case. So I think you need to see if the return of the join is `Any`, and if so check if either `objectType1` or `objectType2` is `Any`. If neither of them is, return nullptr. Otherwise, return `OptionalType::get(unwrappedJoin)`.

Reply:
@rudkx Thanks, fixed!